src/neo/Network/P2P/SyncManager.cs

```diff
@@ -0,0 +1,204 @@
+        private uint lastTaskIndex = 0;
+
+        public SyncManager(NeoSystem system)
+        {
+            this.system = system;
+        }
+
+        protected override void OnReceive(object message)
+        {
+            switch (message)
+            {
+                case Register register:
+                    OnRegister(register.Node);
+                    break;
+                case Block block:
+                    OnReceiveBlock(block);
+                    break;
+                case PersistedBlockIndex blockIndex:
+                    OnReceivePersistedBlockIndex(blockIndex);
+                    break;
+                case InvalidBlockIndex invalidBlockIndex:
+                    OnReceiveInvalidBlockIndex(invalidBlockIndex);
+                    break;
+                case StartSync _:
+                    RequestSync();
+                    break;
+                case Timer _:
+                    OnTimer();
+                    break;
+                case Terminated terminated:
+                    OnTerminated(terminated.ActorRef);
+                    break;
+            }
+        }
+
+        private void OnRegister(RemoteNode node)
+        {
+            Context.Watch(Sender);
+            nodes.Add(Sender, node);
+            RequestSync();
+        }
+
+        private void OnReceiveBlock(Block block)
+        {
+            var node = nodes.Values.FirstOrDefault(p => p.session.IndexTasks.ContainsKey(block.Index));
+            if (node is null) return;
+            node.session.IndexTasks.Remove(block.Index);
+            receivedBlockIndex.TryAdd(block.Index, node);
+            system.Blockchain.Tell(block);
+            RequestSync();
+        }
+
+        private void OnReceivePersistedBlockIndex(PersistedBlockIndex blockIndex)
+        {
+            receivedBlockIndex.Remove(blockIndex.PersistedIndex);
+        }
+
+        private void OnReceiveInvalidBlockIndex(InvalidBlockIndex invalidBlockIndex)
+        {
+            receivedBlockIndex.TryGetValue(invalidBlockIndex.InvalidIndex, out RemoteNode node);
+            if (node is null) return;
+            node.session.InvalidBlockCount++;
+            node.session.IndexTasks.Remove(invalidBlockIndex.InvalidIndex);
+            receivedBlockIndex.Remove(invalidBlockIndex.InvalidIndex);
+            AssignTask(invalidBlockIndex.InvalidIndex, node.session);
+            return;
+        }
+
+        private void RequestSync()
+        {
+            if (GetTasksCount() >= MaxTasksCount || nodes.Count() == 0) return;
+            SendPingMessage();
+            highestBlockIndex = nodes.Values.Max(p => p.LastBlockIndex);
+            if (lastTaskIndex == 0)
+                lastTaskIndex = Blockchain.Singleton.Height;
+            for (int i = 0; i < failedTasks.Count(); i++)
+            {
+                if (failedTasks[i] <= Blockchain.Singleton.Height)
+                {
+                    failedTasks.Remove(failedTasks[i]);
```

Comment:
we can move `failedTasks.Remove(persistedIndex)` to `OnReceivePersistedBlockIndex` method.